src/Logger/Log.php

```diff
@@ -0,0 +1,338 @@
+<?php
+
+namespace Utopia\Logger;
+
+use Exception;
+use Utopia\Logger\Log\Breadcrumb;
+use Utopia\Logger\Log\User;
+
+class Log
+{
+    const TYPE_DEBUG = "debug";
+    const TYPE_ERROR = "error";
+    const TYPE_WARNING = "warning";
+    const TYPE_INFO = "info";
+    const TYPE_VERBOSE = "verbose";
+
+    const ENVIRONMENT_PRODUCTION = "production";
+    const ENVIRONMENT_STAGING = "staging";
+
+    /**
+     * @var float (required, set by default to microtime(true))
+     */
+    protected float $timestamp;
+
+    /**
+     * @var string (required, for example 'Log::TYPE_INFO')
+     */
+    protected string $type;
+
+    /**
+     * @var string (required)
+     */
+    protected string $message;
+
+    /**
+     * @var string (required)
+     */
+    protected string $version;
+
+    /**
+     * @var string (required)
+     */
+    protected string $environment;
+
+    /**
+     * @var string (required)
+     */
+    protected string $action;
+
+    /**
+     * @var array (optional)
+     */
+    protected array $tags = [];
+
+    /**
+     * @var array (optional)
+     */
+    protected array $extra = [];
+
+    /**
+     * @var string (optional)
+     */
+    protected string $namespace = "UNKNOWN";
+
+    /**
+     * @var string|null (optional)
+     */
+    protected ?string $server = null;
+
+    /**
+     * @var User|null (optional)
+     */
+    protected ?User $user = null;
+
+    /**
+     * @var Breadcrumb[] (optional)
+     */
+    protected array $breadcrumbs = [];
+
+    /**
+     * Log constructor.
+     */
+    public function __construct()
+    {
+        $this->timestamp = \microtime(true);
+    }
+
+    /**
+     * Set a type
+     *
+     * @param string $type (required, can be 'log', 'error' or 'warning')
+     * @return void
+     * @throws Exception
+     */
+    public function setType(string $type): void {
+        switch ($type) {
+            case self::TYPE_DEBUG:
+            case self::TYPE_ERROR:
+            case self::TYPE_VERBOSE:
+            case self::TYPE_INFO:
+            case self::TYPE_WARNING:
+                break;
+            default: throw new Exception("Unsupported log type. Must be one of Log::TYPE_DEBUG, Log::TYPE_ERROR, Log::TYPE_WARNING, Log::TYPE_INFO, Log::VERBOSE.");
+        }
+
+        $this->type = $type;
+    }
+
+    /**
+     * Get the type of this log
+     *
+     * @return string (can be 'log', 'error' or 'warning')
+     */
+    public function getType(): string {
```

Comment:
We need to have the opening brace of a function in a new line according to PHP Fig standards. Please fix it here